hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -46,6 +46,23 @@ object InferPType {
     }
   }
 
+  def deepRenameStruct(typeWithNames: Type, targetPType: PType): PType = {
```

Comment:
This should be a method on PType, with a default implementation `return this`, and collection types override (e.g. for PStruct):

```scala
override def deepRename(t: Type) = deepRenameStruct(t.asInstanceOf[TStruct])
def deepRenameStruct(t: TStruct)
```

(PCanonicalBaseStruct implements the `deepRenameStruct`)

You're also missing all the other collections.

Reply:
Sure. It's hard to understand how CastRename is used without seeing test cases. The only test cases that existed for CastRename in IRSuite were for Struct and Array.

Will address these likely after I return.